WordPressAuthenticator/OAuth/ProofKeyForCodeExchange.swift

```diff
@@ -25,21 +25,72 @@ struct ProofKeyForCodeExchange {
         }
     }
 
-    let codeVerifier: String
+    let codeVerifier: CodeVerifier
     let method: Method
 
-    init(codeVerifier: String, method: Method) {
+    init(codeVerifier: CodeVerifier = CodeVerifier(), method: Method = .s256) {
         self.codeVerifier = codeVerifier
         self.method = method
     }
 
     var codeCallenge: String {
         switch method {
         case .s256:
-            // TODO: code_challenge = BASE64URL-ENCODE(SHA256(ASCII(code_verifier)))
-            fatalError()
+            // The spec defines code_challenge for the s256 mode as:
+            //
+            // code_challenge = BASE64URL-ENCODE(SHA256(ASCII(code_verifier)))
+            //
+            // We don't need the ASCII conversion, because we build `CodeVerifier` from URL safe
+            // characters.
+            //
+            // In the same way, it is safe to force unwrap the `Data` conversion because the
+            // `CodeVerifier` input is guaranteed to have only UTF-8 representable characters.
+            //
+            // Also worth noting that `.data(using: .utf8)` cannot technically return `nil` anyway.
+            // See https://forums.swift.org/t/can-encoding-string-to-data-with-utf8-fail/22437/4.
+            let rawData = codeVerifier.rawValue.data(using: .utf8)!
+            let hashedData: Data = rawData.sha256Hashed()
+            return hashedData.base64URLEncodedString()
         case .plain:
-            return codeVerifier
+            return codeVerifier.rawValue
         }
     }
 }
+
+extension ProofKeyForCodeExchange {
+
+    struct CodeVerifier: Equatable {
+
+        let rawValue: String
+
+        // From the docs: using the unreserved characters [A-Z] / [a-z] / [0-9] / "-" / "." / "_" / "~"
+        // That is, URL safe characters.
+        //
+        // Notice that Swift offers `CharacterSet.urlQueryAllowed` to represent this set of characters.
+        // However, there is no straightforward way to convert a `CharacterSet` to a `Set<Character>`.
+        // See for example https://nshipster.com/characterset/.
+        private let allowedCharacters = Set("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-._~")
```

Comment:
Would it be work yoinking this into an extension somewhere?

Reply:
Thank you for the suggestion. Addressed in e23de84e